pkg/provider/provider.go

```diff
@@ -272,82 +288,119 @@ func (p *Provider) buildServiceAndRoutersForTrafficSplit(t *topology.Topology, c
+	tcpRule := buildTCPRouterRule()
 
-				rule = buildHTTPRuleFromTrafficSplitIndirect(tsSvc)
-				rtrMiddlewaresindirect := addToSliceCopy(middlewares, whitelistIndirectKey)
+	for _, svcPort := range tsSvc.Ports {
+		backendSvcs := make([]dynamic.TCPWRRService, len(ts.Backends))
 
-				indirectRtrKey := getRouterKeyFromTrafficSplitIndirect(ts, svcPort.Port)
-				cfg.HTTP.Routers[indirectRtrKey] = buildHTTPRouter(rule, entrypoint, rtrMiddlewaresindirect, svcKey, priorityTrafficTargetIndirect)
+		for i, backend := range ts.Backends {
+			backendSvcKey := getServiceKeyFromTrafficSplitBackend(ts, svcPort.Port, backend)
+			cfg.TCP.Services[backendSvcKey] = buildTCPSplitTrafficBackendService(backend, svcPort.TargetPort.IntVal)
+			backendSvcs[i] = dynamic.TCPWRRService{
+				Name:   backendSvcKey,
+				Weight: getIntRef(backend.Weight),
 			}
 		}
-	case k8s.ServiceTypeTCP:
-		tcpRule := buildTCPRouterRule()
-
-		for _, svcPort := range tsSvc.Ports {
-			backendSvcs := make([]dynamic.TCPWRRService, len(ts.Backends))
-
-			for i, backend := range ts.Backends {
-				backendSvcKey := getServiceKeyFromTrafficSplitBackend(ts, svcPort.Port, backend)
-				cfg.TCP.Services[backendSvcKey] = buildTCPSplitTrafficBackendService(backend, svcPort.TargetPort.IntVal)
-				backendSvcs[i] = dynamic.TCPWRRService{
-					Name:   backendSvcKey,
-					Weight: getIntRef(backend.Weight),
-				}
-			}
 
-			entrypoint, err := p.buildTCPEntrypoint(tsSvc, svcPort.Port)
-			if err != nil {
-				p.logger.Errorf("Unable to build TCP entrypoint for TrafficTarget %q and port %d: %v", tsKey, svcPort.Port, err)
-				continue
+		entrypoint, err := p.buildTCPEntrypoint(tsSvc, svcPort.Port)
+		if err != nil {
+			p.logger.Errorf("Unable to build TCP entrypoint for TrafficTarget %q and port %d: %v", tsKey, svcPort.Port, err)
+			continue
+		}
+
+		key := getServiceRouterKeyFromService(tsSvc, svcPort.Port)
+		cfg.TCP.Services[key] = buildTCPServiceFromTrafficSplit(backendSvcs)
+		cfg.TCP.Routers[key] = buildTCPRouter(tcpRule, entrypoint, key)
+	}
+}
+
+func (p *Provider) buildUDPServiceAndRoutersForTrafficSplit(cfg *dynamic.Configuration, tsKey topology.Key, ts *topology.TrafficSplit, tsSvc *topology.Service) {
+	for _, svcPort := range tsSvc.Ports {
+		backendSvcs := make([]dynamic.UDPWRRService, len(ts.Backends))
+
+		for i, backend := range ts.Backends {
+			backendSvcKey := getServiceKeyFromTrafficSplitBackend(ts, svcPort.Port, backend)
+			cfg.UDP.Services[backendSvcKey] = buildUDPSplitTrafficBackendService(backend, svcPort.TargetPort.IntVal)
+			backendSvcs[i] = dynamic.UDPWRRService{
+				Name:   backendSvcKey,
+				Weight: getIntRef(backend.Weight),
 			}
+		}
 
-			key := getServiceRouterKeyFromService(tsSvc, svcPort.Port)
-			cfg.TCP.Services[key] = buildTCPServiceFromTrafficSplit(backendSvcs)
-			cfg.TCP.Routers[key] = buildTCPRouter(tcpRule, entrypoint, key)
+		entrypoint, err := p.buildUDPEntrypoint(tsSvc, svcPort.Port)
```

Comment:
Same issue here, backend services will be created even for TCP ports.

Reply:
See above.